### PR TITLE
refactor(node): avoid mutation upon persisting previous chain state

### PR DESCRIPTION
**Pull request by aesedepece**
Co-authored-by: Mario Cao <mariocao@gmail.com>

### CODE REVIEW

File: node/src/actors/chain_manager/mod.rs [hunk 1]
```diff
@@ -245,42 +245,47 @@ impl SystemService for ChainManager {}
...
 /// Auxiliary methods for ChainManager actor
 impl ChainManager {
-    /// Method to persist backup of chain_state into storage
-    fn persist_last_chain_state(&self, ctx: &mut Context<Self>) {
-        match self.last_chain_state.chain_info.as_ref() {
-            Some(x) => x,
+    /// Persist previous chain state into storage
+    fn persist_previous_chain_state(&mut self, ctx: &mut Context<Self>) {
+        // It makes no sense to persist empty chain states
+        match self.previous_chain_state.chain_info {
```
Comment (lrubiorod) [inline node/src/actors/chain_manager/mod.rs:251]:
```suggestion
        if self.previous_chain_state.chain_info.is_none() {
```

Comment (aesedepece) [reply node/src/actors/chain_manager/mod.rs]: Nice catch! Changed!

File: node/src/actors/chain_manager/actor.rs [hunk 2]
```diff
@@ -246,8 +246,8 @@ impl ChainManager {
                 });
 
                 // Persist current chain state
-                act.last_chain_state = act.chain_state.clone();
-                act.persist_chain_state(ctx);
+                act.previous_chain_state = act.chain_state.clone();
+                act.persist_previous_chain_state(ctx);
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/actor.rs:250]: This is repeated twice, let's move the clone inside a method:

```suggestion
                act.persist_current_chain_state(ctx);
```

File: node/src/actors/chain_manager/mod.rs [hunk 3]
```diff
@@ -1152,30 +1154,36 @@ impl ChainManager {
...
+                        log::debug!("Before update: superblock {:?}", superblock_beacon);
+
+                        // Take last beacon from superblock state and use it in current chain_info
                         act.chain_state.chain_info.as_mut().unwrap().highest_superblock_checkpoint =
                             act.chain_state.superblock_state.get_beacon();
+
                         if act.sm_state == StateMachine::Synced {
                             // Persist previous_chain_state with current superblock_state
-                            act.persist_chain_state(ctx);
+                            act.persist_previous_chain_state(ctx);
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mod.rs:1170]: Here we call `persist_previous_chain_state` but we don't update `self.previous_state`. This means that consecutive calls to `persist_previous_chain_state` will all persist the same state (the initial one). We must update `self.previous_chain_state` after persisting:

```suggestion
                            act.persist_previous_chain_state(ctx);
                            act.previous_chain_state = act.chain_state.clone();
```

Not sure if we should also add this line to the other places, where `self.previous_chain_state` is updated before.

File: node/src/actors/chain_manager/actor.rs (hunk 2 above)
Comment (aesedepece) [reply node/src/actors/chain_manager/actor.rs]: Done, it's `self.advance_chain_state` now

File: node/src/actors/chain_manager/mod.rs (hunk 3 above)
Comment (aesedepece) [reply node/src/actors/chain_manager/mod.rs:1175]: Intredasting. @mariocao what do you think?

**Comment by tmpolaczyk** [comment]
We could simplify the logic by giving names to the old copies of the chain state. For example when we know that there is consensus for superblock n we want to persist the chain state at superblock n to the storage. And when we are synchronizing up to superblock m, we want to persist the chain state at superblock m. The only problem is that the current chain state is n+2. 

So if instead of `previous_chain_state: ChainState` we use `previous_chain_states: HashMap<SuperEpoch, ChainState>`, then the logic becomes straight forward because we can just call `self.persist_chain_state(consensus_superepoch)`.

File: node/src/actors/chain_manager/mod.rs (hunk 3 above)
Comment (mariocao) [reply node/src/actors/chain_manager/mod.rs:1175]: seems correct...

File: node/src/actors/chain_manager/mod.rs [hunk 4]
```diff
@@ -1034,10 +1041,10 @@ impl ChainManager {
...
                             act.get_superblock_beacon()
                         );
-                        // Persist current_chain_state with
-                        // current_superblock_state and the new superblock beacon
-                        act.last_chain_state = act.chain_state.clone();
-                        act.persist_chain_state(ctx);
+
+                        // Copy current chain state into previous chain state, and persist it
+                        act.advance_chain_state();
+                        act.persist_previous_chain_state(ctx);
```
Comment (lrubiorod) [inline node/src/actors/chain_manager/mod.rs:1047]: Here you are moving current_chain_state to previous and persist previous, so... you are persisting the current chain_state. I think it should be first persist the previous one and update the new one as previous
```suggestion
                        act.persist_previous_chain_state(ctx);
                        act.advance_chain_state();
```

File: node/src/actors/chain_manager/mod.rs [hunk 5]
```diff
@@ -1152,30 +1159,36 @@ impl ChainManager {
...
+                        log::debug!("Before update: superblock {:?}", superblock_beacon);
+
+                        // Take last beacon from superblock state and use it in current chain_info
                         act.chain_state.chain_info.as_mut().unwrap().highest_superblock_checkpoint =
                             act.chain_state.superblock_state.get_beacon();
+
                         if act.sm_state == StateMachine::Synced {
                             // Persist previous_chain_state with current superblock_state
-                            act.persist_chain_state(ctx);
+                            act.persist_previous_chain_state(ctx);
```
Comment (lrubiorod) [inline node/src/actors/chain_manager/mod.rs:1175]:
```suggestion
                            act.persist_previous_chain_state(ctx);
                            act.advance_chain_state();
```
I agree

Comment (aesedepece) [reply node/src/actors/chain_manager/mod.rs:1228]: Done!

File: node/src/actors/chain_manager/mod.rs (hunk 3 above)
Comment (aesedepece) [reply node/src/actors/chain_manager/mod.rs:1228]: Done!

**Review by tmpolaczyk — APPROVED** [review]
I found an edge case where the storage is not persisted, but will open a separate PR for that

**Review by lrubiorod — APPROVED** [review]

**Pull request merged**